core/chains/evm/log/registrations.go

```diff
@@ -93,13 +98,31 @@ func (r *registrations) addSubscriber(sub *subscriber) (needsResubscribe bool) {
 	return
 }
 
-func (r *registrations) removeSubscriber(sub *subscriber) (needsResubscribe bool) {
+// checkAddSubscriber makes sure we aren't violating any assumptions by adding
+// this subscriber
+func (r *registrations) checkAddSubscriber(sub *subscriber) {
+	if sub.opts.MinIncomingConfirmations <= 0 {
+		r.logger.Panicw(fmt.Sprintf("LogBroadcaster requires that MinIncomingConfirmations must be at least 1 (got %v). Logs must have been confirmed in at least 1 block, it does not support reading logs from the mempool before they have been mined.", sub.opts.MinIncomingConfirmations), "addr", sub.opts.Contract.Hex(), "jobID", sub.listener.JobID())
+	}
+
 	jobID := sub.listener.JobID()
-	if _, exists := r.registeredSubs[sub]; !exists {
-		r.logger.Panicf("Cannot remove subscriber: subscription %p with job ID %v is not registered", sub, jobID)
+	if _, exists := r.registeredSubs[sub]; exists {
+		r.logger.Panicf("Cannot add subscriber %p for job ID %v: already added", sub, jobID)
 	}
-	delete(r.registeredSubs, sub)
-	r.logger.Tracef("Added subscription %p with job ID %v", sub, jobID)
+	r.registeredSubs[sub] = struct{}{}
+	addrs, exists := r.jobIDAddrs[jobID]
+	if !exists {
+		r.jobIDAddrs[jobID] = make(map[common.Address]struct{})
+	}
+	if _, exists := addrs[sub.opts.Contract]; exists {
+		r.logger.Panicf("Cannot add subscriber %p: only one subscription is allowed per jobID/contract address. There is already a subscription with job ID %v listening on %s", sub, jobID, sub.opts.Contract.Hex())
+	}
+	r.jobIDAddrs[jobID][sub.opts.Contract] = struct{}{}
+}
+
+func (r *registrations) removeSubscriber(sub *subscriber) (needsResubscribe bool) {
+	r.checkRemoveSubscriber(sub)
```

Comment:
Would it be a little cleaner to return an err here and then panic from this one place, rather than three similar inline panics? Leaves `checkRemoveSubscriber` more pure and testable that way.